service/gcs/main.go

```diff
@@ -5,21 +5,60 @@ import (
 	"fmt"
 	"io"
 	"os"
+	"path/filepath"
 	"syscall"
 	"time"
 
+	"github.com/Microsoft/opengcs/internal/cgrouputils"
 	"github.com/Microsoft/opengcs/internal/oc"
 	"github.com/Microsoft/opengcs/internal/runtime/hcsv2"
 	"github.com/Microsoft/opengcs/service/gcs/bridge"
 	"github.com/Microsoft/opengcs/service/gcs/core/gcs"
 	"github.com/Microsoft/opengcs/service/gcs/runtime/runc"
 	"github.com/Microsoft/opengcs/service/gcs/transport"
 	"github.com/containerd/cgroups"
+	v1 "github.com/containerd/cgroups/stats/v1"
 	oci "github.com/opencontainers/runtime-spec/specs-go"
 	"github.com/sirupsen/logrus"
 	"go.opencensus.io/trace"
 )
 
+func memoryLogFormat(metrics *v1.Metrics) logrus.Fields {
+	return logrus.Fields{
+		"memory-usage":       metrics.Memory.Usage.Usage,
+		"memory-usage-max":   metrics.Memory.Usage.Max,
+		"memory-usage-limit": metrics.Memory.Usage.Limit,
+		"swap-usage":         metrics.Memory.Swap.Usage,
+		"swap-usage-max":     metrics.Memory.Swap.Max,
+		"swap-usage-limit":   metrics.Memory.Swap.Limit,
+		"kernel-usage":       metrics.Memory.Kernel.Usage,
+		"kernel-usage-max":   metrics.Memory.Kernel.Max,
+		"kernel-usage-limit": metrics.Memory.Kernel.Limit,
+	}
+}
+
+func readMemoryEvents(efdFile *os.File, cgName string, threshold int64, cg cgroups.Cgroup) {
+	// Buffer must be >= 8 bytes for eventfd reads
+	// http://man7.org/linux/man-pages/man2/eventfd.2.html
+	buf := make([]byte, 8)
+	for {
+		_, err := efdFile.Read(buf)
+		if err != nil {
+			logrus.WithError(err).Warn("opengcs::readEvents - failed to read from eventfd")
+			return
+		}
+		msg := fmt.Sprintf("opengcs::readMemoryEvents - memory usage for cgroup '%s' exceeded '%d' byte threshold", cgName, threshold)
+		time.Sleep(time.Second)
```

Comment:
What happens if we don't sleep here? Are the values noticeably different?

Reply:
For the gcs case yes as the event gets fired and theres no action taken besides logging so the numbers can continue to change. For the oom case not really.

Reply:
In that case, why do we sleep? Don't we want to know what these values are at the time we hit the threshold?

Reply:
I suggested this originally. My intuition was that it might be helpful to wait a second in case there are a series of allocations. I am guessing that the "max" fields will be the high water mark, in which case we don't have to worry about what the memory state is at the precise time we exceed the threshold.

Reply:
max is the highwater mark correct, usage is the snapshot at the current point in time. On event notification if this is the first time we've exceeded the threshold Usage and Max would be very similar (depending on how long Stat takes to run. Stat calls the corresponding Stat method on all subsystems not just memory).

Reply:
That seems reasonable. My request then is we add a comment here explaining why we are sleeping (i.e. in case there is a series of allocations).